lms/static/scripts/frontend_apps/components/ContentSelector.js

```diff
@@ -0,0 +1,186 @@
+ *
+ * @typedef ErrorInfo
+ * @prop {string} title
+ * @prop {Error} error
+ *
+ * @typedef {'lms'|'url'|null} DialogType
+ *
+ * @typedef {import('./FilePickerApp').Content} Content
+ */
+
+/**
+ * @typedef ContentSelectorProps
+ * @prop {(ei: ErrorInfo) => void} onError
+ * @prop {(c: Content) => void} onSelectContent
+ * @prop {DialogType} [defaultActiveDialog]
+ */
+
+/**
+ * Component that allows the user to choose the content for an assignment.
+ *
+ * @param {ContentSelectorProps} props
+ */
+export default function ContentSelector({
+  defaultActiveDialog = null,
+  onError,
+  onSelectContent,
+}) {
+  const {
+    api: { authToken },
+    filePicker: {
+      canvas: { enabled: canvasEnabled, listFiles: listFilesApi },
+      google: {
+        clientId: googleClientId,
+        developerKey: googleDeveloperKey,
+        origin: googleOrigin,
+      },
+      vitalSource: { enabled: vitalSourceEnabled },
+    },
+  } = useContext(Config);
+
+  const [isLoadingIndicatorVisible, setLoadingIndicatorVisible] =
+    useState(false);
+  const [activeDialog, setActiveDialog] = useState(
+    /** @type {DialogType} */ (defaultActiveDialog)
+  );
+
+  const cancelDialog = () => {
+    setLoadingIndicatorVisible(false);
+    setActiveDialog(null);
+  };
+
+  /** @param {DialogType} type */
+  const selectDialog = type => {
+    setLoadingIndicatorVisible(true);
+    setActiveDialog(type);
+  };
+  // Initialize the Google Picker client if credentials have been provided.
+  // We do this eagerly to make the picker load faster if the user does click
+  // on the "Select from Google Drive" button.
+  const googlePicker = useMemo(() => {
+    if (!googleClientId || !googleDeveloperKey || !googleOrigin) {
+      return null;
+    }
+    return new GooglePickerClient({
+      developerKey: googleDeveloperKey,
+      clientId: googleClientId,
+
+      // If the form is being displayed inside an iframe, then the backend
+      // must provide the URL of the top-level frame to us so we can pass it
+      // to the Google Picker API. Otherwise we can use the URL of the current
+      // tab.
+      origin: window === window.top ? window.location.href : googleOrigin,
+    });
+  }, [googleDeveloperKey, googleClientId, googleOrigin]);
+
+  /** @param {string} url */
+  const selectURL = url => {
+    cancelDialog();
+    onSelectContent({ type: 'url', url });
+  };
+
+  /** @param {File} file */
+  const selectLMSFile = file => {
+    cancelDialog();
+    onSelectContent({ type: 'file', file });
+  };
+
+  let dialog;
+  switch (activeDialog) {
+    case 'url':
+      dialog = <URLPicker onCancel={cancelDialog} onSelectURL={selectURL} />;
+      break;
+    case 'lms':
+      dialog = (
+        <LMSFilePicker
+          authToken={authToken}
+          listFilesApi={listFilesApi}
+          onCancel={cancelDialog}
+          onSelectFile={selectLMSFile}
+        />
+      );
+      break;
+    default:
+      dialog = null;
+  }
```

Comment:
While this is a little more compact, I prefer using a falsey value to represent "no dialog shown" because you can treat the value as a boolean in cases where you want to test "is this value present"? (eg. if you wanted to render something whenever any or no dialog is shown).